# Chapitre 6 - Différentes utilisations dumachine learning avec Python (4ème partie)

## 6.7 Le deep learning avec Keras

### 6.7.2 Principe d’un réseau de neurone et première utilisation avec Keras

Le package Keras que nous avons installé est basé sur une structure extrêmement
simple à mettre en oeuvre. On commence par définir le type de réseau utilisé. Dans
cet ouvrage, il s’agira de réseaux séquentiels.

Si nous prenons les données sur les télécommunications que nous avons utilisées
précédemment, nous pouvons simplement utiliser Keras pour créer un réseau de neurones assez évolué (on suppose ici que les données ont été préparées comme
dans la première partie de ce chapitre avec Scikit-Learn) :

In [1]:
import pandas as pd
import numpy as np
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.layers import BatchNormalization, Activation

In [2]:
x=pd.read_csv("../data/x_telecom.csv",index_col=0)
y=pd.read_csv("../data/y_telecom.csv",index_col=0)

In [3]:
from sklearn.model_selection import train_test_split
# on sépare nos données en apprentissage/validation avec Scikit-Learn
x_train, x_test, y_train, y_test=train_test_split(x,y,test_size=0.2)

In [4]:
# on crée le modèle
model = Sequential()

In [5]:
# on ajoute des couches avec des fonctions d’activation
model.add(Dense(50, input_shape=(21,)))
model.add(Activation("relu"))
# cette couche permet de normaliser les données
model.add(BatchNormalization())
# cette couche va faire en sorte d’éviter l’overfitting
model.add(Dropout(0.1))
# on ajoute un autre groupe de couches
model.add(Dense(10))
model.add(Activation("relu"))
model.add(BatchNormalization())
model.add(Dropout(0.1))
# on ajoute la couche de sortie
model.add(Dense(1))
model.add(Activation('sigmoid'))


In [6]:
# on définit la fonction de perte et la fonction d’optiimsation
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=["accuracy"])

In [7]:
# on ajuste le modèle
model.fit(x_train, y_train, batch_size=100, epochs=100)

Epoch 1/100
27/27 [==============================] - 0s 2ms/step - loss: 0.8044 - accuracy: 0.5949
Epoch 2/100
27/27 [==============================] - 0s 2ms/step - loss: 0.6631 - accuracy: 0.6718
Epoch 3/100
27/27 [==============================] - 0s 2ms/step - loss: 0.5822 - accuracy: 0.7198
Epoch 4/100
27/27 [==============================] - ETA: 0s - loss: 0.5760 - accuracy: 0.69 - 0s 2ms/step - loss: 0.5351 - accuracy: 0.7663
Epoch 5/100
27/27 [==============================] - 0s 2ms/step - loss: 0.4934 - accuracy: 0.7926
Epoch 6/100
27/27 [==============================] - 0s 2ms/step - loss: 0.4471 - accuracy: 0.8342
Epoch 7/100
27/27 [==============================] - 0s 2ms/step - loss: 0.4148 - accuracy: 0.8530
Epoch 8/100
27/27 [==============================] - 0s 2ms/step - loss: 0.3876 - accuracy: 0.8616
Epoch 9/100
27/27 [==============================] - 0s 2ms/step - loss: 0.3498 - accuracy: 0.8845
Epoch 10/100
27/27 [==============================] - 0s 2ms/step -

27/27 [==============================] - 0s 2ms/step - loss: 0.1270 - accuracy: 0.9572
Epoch 83/100
27/27 [==============================] - 0s 2ms/step - loss: 0.1179 - accuracy: 0.9580
Epoch 84/100
27/27 [==============================] - 0s 2ms/step - loss: 0.1168 - accuracy: 0.9625
Epoch 85/100
27/27 [==============================] - 0s 2ms/step - loss: 0.1219 - accuracy: 0.9587
Epoch 86/100
27/27 [==============================] - 0s 2ms/step - loss: 0.1112 - accuracy: 0.9617
Epoch 87/100
27/27 [==============================] - 0s 2ms/step - loss: 0.1259 - accuracy: 0.9580
Epoch 88/100
27/27 [==============================] - 0s 2ms/step - loss: 0.1210 - accuracy: 0.9640
Epoch 89/100
27/27 [==============================] - 0s 2ms/step - loss: 0.1275 - accuracy: 0.9599
Epoch 90/100
27/27 [==============================] - 0s 2ms/step - loss: 0.1183 - accuracy: 0.9621
Epoch 91/100
27/27 [==============================] - 0s 2ms/step - loss: 0.1200 - accuracy: 0.9602
Epoch 92/100


In [8]:
model.predict_proba(x_test)

Instructions for updating:
Please use `model.predict()` instead.


array([[9.97193336e-01],
       [3.62157822e-02],
       [9.85870242e-01],
       [1.70528471e-01],
       [3.50208282e-02],
       [7.84272850e-02],
       [7.66124427e-02],
       [1.00000000e+00],
       [2.41908431e-03],
       [1.49467885e-02],
       [2.03776360e-03],
       [8.20812583e-03],
       [1.00000000e+00],
       [5.10457158e-03],
       [4.07670468e-01],
       [3.23989987e-03],
       [7.64716864e-02],
       [6.17393315e-01],
       [2.71230936e-04],
       [3.12215984e-02],
       [2.53165126e-01],
       [9.21270132e-01],
       [4.37194109e-03],
       [9.41532850e-03],
       [1.15052462e-02],
       [9.84957337e-01],
       [3.87486815e-03],
       [1.29806399e-02],
       [1.67300701e-02],
       [1.74522400e-04],
       [1.07380748e-03],
       [2.69854069e-03],
       [1.71230733e-01],
       [6.13301992e-04],
       [6.64949417e-03],
       [2.68164575e-02],
       [1.21217966e-02],
       [9.99880195e-01],
       [3.58045101e-03],
       [1.15264654e-02],


In [9]:
# on calcule l’AUC pour les données de validation
from sklearn.metrics import roc_auc_score
print("AUC pour RN :", roc_auc_score(y_test,
                                     model.predict_proba(x_test).reshape(-1)))

AUC pour RN : 0.9186596668611055


Les réseaux de neurones sont des méthodes extrêmement efficaces à condition
de bien les paramétrer ce qui peut s‘avérer très difficile.

### 6.7.3 Le deep learning et les réseaux de neurones à convolutions

On va maintenant travailler sur des images. On prendra les données fashion-mnist :

In [10]:
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, Activation, Flatten
from tensorflow.keras.layers import Conv2D, MaxPool2D
from tensorflow.keras.datasets import fashion_mnist

In [11]:
# récupération des données
(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()

# les données sont transformées pour passer sous un format d’image
# standard avec des valeurs entre 0 et 1
x_train = (x_train/255).reshape(x_train.shape[0], 28, 28,1)
x_test = (x_test/255).reshape(x_test.shape[0], 28, 28,1)

# passage à 10 colonnes pour y(une par modalité)
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

In [12]:
# construction du modèle
model = Sequential()
model.add(Conv2D(32, (3, 3), activation = 'relu', input_shape = (28,28,1)))
model.add(Conv2D(32, (3, 3), activation = 'relu'))
model.add(MaxPool2D(pool_size=(2,2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(10, activation='softmax'))

In [13]:
# compilation du modèle
model.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

In [14]:
# ajustement du modèle
model.fit(x_train, y_train, batch_size=100, epochs=10, verbose=1)

Epoch 1/10
600/600 [==============================] - 43s 72ms/step - loss: 0.5484 - accuracy: 0.8068
Epoch 2/10
600/600 [==============================] - 44s 73ms/step - loss: 0.3659 - accuracy: 0.8703
Epoch 3/10
600/600 [==============================] - 43s 71ms/step - loss: 0.3183 - accuracy: 0.8852
Epoch 4/10
600/600 [==============================] - 46s 76ms/step - loss: 0.2926 - accuracy: 0.8932
Epoch 5/10
600/600 [==============================] - 46s 77ms/step - loss: 0.2670 - accuracy: 0.9036
Epoch 6/10
600/600 [==============================] - 51s 85ms/step - loss: 0.2470 - accuracy: 0.9101
Epoch 7/10
600/600 [==============================] - 48s 81ms/step - loss: 0.2328 - accuracy: 0.9141
Epoch 8/10
600/600 [==============================] - 50s 84ms/step - loss: 0.2179 - accuracy: 0.9205
Epoch 9/10
600/600 [==============================] - 51s 85ms/step - loss: 0.2085 - accuracy: 0.9229
Epoch 10/10
600/600 [==============================] - 49s 81ms/step - loss: 0.196

In [15]:
# evaluation du modèle
score = model.evaluate(x_test, y_test, verbose=0)

Nous avons appliqué deux couches de convolutions à 32 neurones avec une taille
de fenêtre de 3 par 3. Puis une couche de pooling avec une réduction sur des fenêtres
de 2 par 2. Nous obtenons les résultats suivants :

In [16]:
# obtention de l’accuracy sur les données de validation
print("Pourcentage de bien classées :", score[1])

Pourcentage de bien classées : 0.9215999841690063


On voit qu’on est au-delà de 92 % bien classés sur l’échantillon de validation.
Dans notre cas, on a utilisé des images directement dans le package Keras.

Généralement, les images sont stockées dans des répertoires différents. Keras
possède de nombreux outils pour charger ces images, on peut par exemple utiliser
ImageDataGenerator :

In [17]:
from tensorflow.keras.preprocessing.image import ImageDataGenerator

# on définit un générateur qui pourrait modifier nos images
# (dans ce cas on passe en float 0 à 1)
train_datagen = ImageDataGenerator(rescale=1./255)
test_datagen = ImageDataGenerator(rescale=1./255)

# on définit le dossier dans lequel se trouve les images
train_generator = train_datagen.flow_from_directory('../data/',target_size=(150, 150),
                                                    batch_size=32,class_mode='binary')
# on définit le dossier dans lequel se trouve les images
validation_generator = test_datagen.flow_from_directory('../data/',
                                                        target_size=(150, 150),
                                                        batch_size=32,
                                                        class_mode='binary')

Found 0 images belonging to 0 classes.
Found 0 images belonging to 0 classes.


Vous avez maintenant tous les outils pour vous lancer dans la construction de
réseaux de deep learning avec Python.

### 6.7.4 Aller plus loin : génération de features, transfer learning, RN, GAN …
On charge facilement un modèle avec :

In [18]:
from tensorflow.keras.applications.resnet50 import ResNet50
from tensorflow.keras.preprocessing import image
from tensorflow.keras.applications.resnet50 import preprocess_input
from tensorflow.keras.applications.resnet50 import decode_predictions
import numpy as np

In [19]:
# on utilise le modèle ResNet50
model = ResNet50(weights='imagenet')

In [20]:
# on récupère une nouvelle image
img_path = '../data/elephant.png'
img = image.load_img(img_path, target_size=(224, 224))
x = image.img_to_array(img)
x = np.expand_dims(x, axis=0)
x = preprocess_input(x)

In [21]:
# on applique le modèle
preds = model.predict(x)
print('Prédit (classe/ %):', decode_predictions(preds, top=1)[0])

Prédit (classe/ %): [('n02504458', 'African_elephant', 0.6474536)]


On obtient donc un résultat à partir du modèle chargé. On pourra utiliser ce modèle
et ses poids dans des couches intermédiaires de notre modèle.

Dans ce cas, on a chargé un modèle et on lui a fourni une photographie d’éléphant.
Le modèle considère que c’est un éléphant d’Afrique à 87%.

Nous avons effectué un panorama approfondi de l’utilisation de Python
pour faire du machine learning, du deep learning et du traitement des
données textuelles. Python ressort comme un langage adapté à toutes
ces approches et un outil indispensable pour le data scientist.